database/txn/transaction.go

```diff
@@ -168,6 +201,25 @@ func (t *TransactionRunner) Retry(ctx context.Context, fn func() error) error {
 	return t.retryStrategy(ctx, fn)
 }
 
+// run will execute the input function if there is a semaphore slot available.
+func (t *TransactionRunner) run(ctx context.Context, fn func(context.Context) error) error {
+	if err := t.semaphore.Acquire(ctx, 1); err != nil {
```

Comment:
I'm a bit confused, are we limiting the number of concurrent txns on the juju process based on the available GOMAXPROCS? but this could have unexpected consequences right? I mean dqlite must be handling on their own way how to deal with a growing number of txns so adding an extra control on our side might be limiting on cases we shouldn't be limiting

Reply:
This isn't the way it works unfortunately. As a transaction is a full DB lock starting a new transaction and doing an operation like exec or query will result in error to the effect of DB locked.

The retry mechanism we have deals with but under lots of load and models the retry mechanism will also fail as it reached the maximum amount of retries. The retry strategy also introduces a starvation problem between the go routines.

There is a speed benefit we have proved by having a certain number of retry able transactions in flight but over a certain limit it starts to become a determent to performance.

The solution thus far is to use some kind of ordered write mechanism like a semaphore. The semaphore is a little heavy but as heavy as having 100 retry-able transactions spin locking.

Reply:
sqlite does have a mechanism called BEGIN CONCURRENT that may be of better use to us but the drivers don't make use of it yet. Need more investigation and maybe a PR down stream to see if this will work better for us.

Reply:
> but this could have unexpected consequences right?

So the aim is to force the backing off of transactions to prevent a thundering herd. If you can't write a transaction in 25 seconds then you're likely never will. By backing-off, we're actively forcing the call site of the transaction to enforce back pressure. For example, if we're getting a lot of API requests attempting to write into the database at any one time, then we can tell the API caller to retry at a later date.

> I mean dqlite must be handling on their own way how to deal with a growing number of txns

No. I'm not aware of a database turning away too many transactions, at least sqlite, Postgres, and mongo. There might be a case of too many clients or connections, but not transactions. A transaction will generally time out because a client driver didn't get a response, or in Go's instance we might get a context deadline exceeded.

The problem is this, we use retries for all transactions (see transaction runner) to essentially brute force a transaction to be applied. If the transaction is retriable we keep going. Things like "database is locked" or "checkpoint in progress" (see `IsErrRetryable`) allow us to assume that the next application will work, or subsequent one will et al. If we attempt to use `SetMaxOpenConns` then we actually run into this problem more as [sqlite only allows one write transaction at a time](https://www.sqlite.org/lang_transaction.html).

> SQLite supports multiple simultaneous read transactions coming from separate database connections, possibly in separate threads or processes, but only one simultaneous write transaction

As we're a mixed read and write workload we encounter this a lot. So it's actually simpler and a LOT more efficient to just retry and see what sticks.

The problem that this PR is solving is that at scale (around 400 models with distributed read and write patterns) we end up thrashing the CPU and making no progress. Juju is inherently faster than dqlite can append to the raft log and ship the pages. For that reason, we can limit the number of active transactions to the number of GOMAXPROCS to the number of CPU threads (we'll need to benchmark on the actual final number, but for now it's this one). Therefore we give a fighting chance to other resources having some CPU time.